# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/kr8457/FlyRank-AI-ML-/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

Lane: Lane 1 — Traffic/Content Performance Prediction

Why: I am choosing Lane 1 because predicting organic traffic growth and decay for existing content helps teams proactively flag underperforming assets. By modeling historical performance patterns, content strategists can prioritize updates before traffic drops significantly, optimizing editorial resource allocation.

In [5]:
# Lane selection declaration
lane = "Lane 1: Traffic/Content Performance Prediction"
print(f"Selected Lane: {lane}")

Selected Lane: Lane 1: Traffic/Content Performance Prediction


## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

Unit of Analysis: A single published content page evaluated over monthly time windows.

Target Output: Predicting continuous monthly clicks and identifying binary traffic decay risk (e.g., >20% traffic drop).

Decision & Action: SEO managers decide which pages receive metadata refreshes, content rewrites, or internal link additions within the next planning cycle.

Cost of a Wrong Call:

False Positive (Predicting decay when traffic is fine): Wastes budget and editorial effort re-optimizing healthy content.

False Negative (Missing a real drop): Leads to undetected organic search decay, resulting in lost search visibility and reduced organic leads.

In [6]:
# Decision framework variables
unit_of_analysis = "page_month"
target_variable = "traffic_decay_risk"
print(f"Unit: {unit_of_analysis} | Target: {target_variable}")

Unit: page_month | Target: traffic_decay_risk


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [15]:
import pandas as pd
import requests
import io
from google.colab import userdata

# Get Hugging Face token from Colab Secrets
hf_token = userdata.get('HF_TOKEN')

# Parquet dataset URL
url = "https://huggingface.co/datasets/FlyRank/internship-warehouse/resolve/main/fact_content_daily_performance_sample.parquet"
headers = {"Authorization": f"Bearer {hf_token}"}

response = requests.get(url, headers=headers)

if response.status_code == 200:
    df = pd.read_parquet(io.BytesIO(response.content))

    print("Dataset Columns:", list(df.columns))
    print(f"Total Daily Sample Records: {len(df):,}")

    # Check for exact column names or fallbacks
    click_col = [c for c in df.columns if 'click' in c.lower()]
    imp_col = [c for c in df.columns if 'impression' in c.lower()]

    if click_col:
        print(f"Average Daily Clicks per Record ({click_col[0]}): {df[click_col[0]].mean():.2f}")
    if imp_col:
        print(f"Average Daily Impressions per Record ({imp_col[0]}): {df[imp_col[0]].mean():.2f}")
else:
    print(f"Failed to fetch dataset. HTTP Status: {response.status_code}")

Dataset Columns: ['report_date', 'client_hash_id', 'content_hash_id', 'client_has_gsc', 'client_has_ga4', 'gsc_data_available', 'ga4_data_available', 'gsc_impressions', 'gsc_clicks', 'gsc_sum_position', 'gsc_avg_position', 'ga4_pageviews', 'ga4_sessions', 'ga4_users', 'ga4_engaged_sessions', 'ga4_total_engagement_sec', 'sessions_organic', 'sessions_direct', 'sessions_referral', 'sessions_social', 'sessions_paid', 'sessions_ai', 'ai_chatgpt', 'ai_perplexity', 'ai_gemini', 'ai_copilot', 'ai_claude', 'ai_meta', 'ai_other', 'scroll_events', 'month']
Total Daily Sample Records: 11,694,072
Average Daily Clicks per Record (gsc_clicks): 0.10
Average Daily Impressions per Record (gsc_impressions): 18.49


Total Daily Performance Sample Records: 11,694,072

Average Daily Clicks per Record (gsc_clicks): 0.10

Average Daily Impressions per Record (gsc_impressions): 18.49

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

What this work CAN claim: This project identifies historical statistical associations and directional trends between historical page performance features and future organic click behavior to serve as a decision-support indicator.

What this work CANNOT claim: This project does NOT establish direct causality, prove internal search engine ranking algorithm mechanisms, or guarantee specific Google ranking positions.

In [16]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
# Scope boundaries validation
claims = ["observed_trends", "decision_support", "directional_correlations"]
non_claims = ["causal_proof", "google_algorithm_prediction"]
print(f"Valid Claims Scope: {claims}")


Valid Claims Scope: ['observed_trends', 'decision_support', 'directional_correlations']


## Self-check
Before you submit, confirm each line honestly:
- [x] Every section above is filled — markdown thinking AND the code that backs it
- [x] The notebook runs top to bottom with no errors (Runtime → Run all)
- [x] No client names, URLs, or private queries anywhere
- [x] My claims use careful words: observed, measured, directional, decision-support
- [x] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.